system/jhtree/jhutil.hpp

```diff
@@ -22,6 +22,106 @@
 #include "jqueue.tpp"
 #include "jhtree.hpp"
 
+//Implementation of a queue using a doubly linked list.  Should possibly move to jlib if it would be generally useful.
+//Currently assumes next and prev fields in the element can be used to maintain the list
+template <class ELEMENT>
+class DListOf
+{
+    typedef DListOf<ELEMENT> SELF;
+    ELEMENT * pHead = nullptr;
+    ELEMENT * pTail = nullptr;
+    unsigned numEntries = 0;
+
+public:
+    void enqueueHead(ELEMENT * element)
+    {
+        assertex(!element->next && !element->prev);
+        if (pHead)
+        {
+            pHead->prev = element;
+            element->next = pHead;
+            pHead = element;
+        }
+        else
+        {
+            pHead = pTail = element;
+        }
+        numEntries++;
+    }
+    void enqueue(ELEMENT * element)
+    {
+        assertex(!element->next && !element->prev);
+        if (pTail)
+        {
+            pTail->next = element;
+            element->prev = pTail;
+            pTail = element;
+        }
+        else
+        {
+            pHead = pTail = element;
+        }
+        numEntries++;
+    }
+    ELEMENT *head() const { return pHead; }
+    ELEMENT *tail() const { return pTail; }
+    void remove(ELEMENT *element)
+    {
+        ELEMENT * next = element->next;
+        ELEMENT * prev = element->prev;
+        assertex(prev || next || element == pHead);
+        if (element == pHead)
+            pHead = next;
+        if (element == pTail)
+            pTail = prev;
+        if (next)
+            next->prev = prev;
+        if (prev)
+            prev->next = next;
+        element->next = nullptr;
+        element->prev = nullptr;
+        numEntries--;
+    }
+    ELEMENT *dequeue()
+    {
+        if (!pHead)
+            return nullptr;
+        ELEMENT * element = pHead;
+        ELEMENT * next = element->next;
+        ELEMENT * prev = nullptr;
```

Comment:
Is this used?

Reply:
I spotted, but failed to push the change.